snapcraft/internal/meta.py

```diff
@@ -116,6 +119,38 @@ def setup_assets(self):
             file_utils.link_or_copy(
                 'gadget.yaml', os.path.join(self.meta_dir, 'gadget.yaml'))
 
+    def write_snap_directory(self):
+        # First migrate the snap directory. It will overwrite any conflicting
+        # files.
+        for root, directories, files in os.walk('snap'):
+            for directory in directories:
+                source = os.path.join(root, directory)
+                destination = os.path.join(self._snap_dir, source)
+                file_utils.create_similar_directory(source, destination)
+
+            for file_path in files:
+                source = os.path.join(root, file_path)
+                destination = os.path.join(self._snap_dir, source)
+                with contextlib.suppress(FileNotFoundError):
+                    os.remove(destination)
+                file_utils.link_or_copy(source, destination)
+
+        # Now write hook wrappers for any hooks included in that directory.
+        self._generate_hook_wrappers()
```

Comment:
> I would only generate hook wrappers for things coming from parts just to not break any assumption a manually crafted hook would have.

If we don't generate wrappers for the manually crafted ones, there's nothing connecting the dots between our magic location (snap/hooks) and snapd's (meta/hooks). Are you thinking we'd just generate a dumb wrapper?

I'm trying to think of assumptions a manually crafted hook might have that would break with our setting environment variables... I'm not coming up with any, do you know? However, I can quickly come up with a breakage if we don't set them: if the manually crafted hook uses any python executables within it, it'll immediately break. And there's no way around it except to move them out of the `snap` directory and install them via a part. The inconsistency there might make for a poor experience.